scipy/sparse/linalg/_isolve/symmlq.py

```diff
@@ -0,0 +1,239 @@
+        https://gitlab.com/petsc/petsc/-/blob/main/src/ksp/ksp/impls/symmlq/symmlq.c
+
+    Examples
+    --------
+    >>> from scipy.sparse import csc_matrix
+    >>> from scipy.sparse.linalg import symmlq
+    >>> A = csc_matrix([[2, -1, 0], [-1, 0.25, -1], [0, -1, 0]], dtype=float)
+    >>> b = np.array([1, 2, 3], dtype=float)
+    >>> x, exitCode = symmlq(A, b)
+    >>> print(exitCode)            # 0 indicates successful convergence
+    0
+    >>> np.allclose(A.dot(x), b)
+    True
+    """
+
+    A, M, x, b, postprocess = make_system(A, M, x0, b)
+
+    n = A.shape[0]
+    if maxiter is None:
+        maxiter = min(10000, n * 10)
+
+    # Check if the R.H.S is a zero vector
+    bnorm = np.linalg.norm(b)
+    if bnorm == 0.:
+        x = np.zeros(n)
+        if show:
+            print("SYMMLQ: Linear solve converged due to zero right-hand side "
+                  "iterations 0")
+        return (postprocess(x), 0)
+
+    # Check initial guess
+    if x0 is None:
+        r = b
+    else:
+        r = b - A.matvec(x)
+
+    if atol is None:
+        atol = tol * bnorm
+    else:
+        atol = max(atol, tol * bnorm)
+
+    c = cold = 1.
+    s = sold = 0.
+    ceta_oold = ceta_old = ceta = 0.
+    w = wbar = vold = uold = np.zeros(n)
+    z = M.matvec(r)
+
+    # Check initial residual norm
+    rnorm = np.linalg.norm(z)  # preconditioned residual norm
+    if rnorm < atol:
+        if callback is not None:
+            callback(rnorm)
+        if show:
+            print("SYMMLQ: Linear solve converged due to reach TOL "
+                  "iterations 0")
+        return (postprocess(x), 0)
+
+    # Check breakdown
+    dp = np.inner(r.conjugate(), z)
+    if abs(dp) < haptol:
+        if callback is not None:
+            callback(rnorm)
+        if show:
+            print("SYMMLQ: Linear solve converged due to HAPPY BREAKDOWN "
+                  "iterations 0")
+        return (postprocess(x), 0)
+
+    if not np.iscomplex(dp) and dp < 0.:
+        if callback is not None:
+            callback(rnorm)
+        if show:
+            print("SYMMLQ: Linear solve not converged due to DIVERGED "
+                  "INDEFINITE PC iterations 0")
+        return (postprocess(x), -1)
+
+    beta = beta1 = np.sqrt(dp)
+    s_prod = abs(beta1)
+    [v, u] = [r, z] / beta
+    wbar = u
+
+    for iter in range(maxiter):
+        # Update
+        if iter > 0:
+            vold, uold = v, u
+            [v, u] = [r, z] / beta
+            w = c * wbar + s * u
+            wbar = - s * wbar + c * u
+            x += ceta * w
+            ceta_oold, ceta_old = ceta_old, ceta
+
+        # Lanczos method
+        r = A.matvec(u)
+        alpha = np.inner(u.conjugate(), r)
+        z = M.matvec(r)
+        r = r - alpha * v - beta * vold
+        z = z - alpha * u - beta * uold
+        betaold = beta
+        dp = np.inner(r.conjugate(), z)
+        # Check breakdown
+        if abs(dp) < haptol:
```

Comment:
This is almost impossible to satisfy reliably for most data since dp is coming from 64-bit float precision. Hence this branch is not accessible. Take two orthogonal noninteger vectors long enough and check the dot product result.